File: Array/560_subarray_sum_equals_k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> hashmap;

        int curr_sum = 0, count = 0;
        for(int num: nums) {
            curr_sum += num;

            if(curr_sum == k) count += 1;
            if(hashmap.find(curr_sum - k) != hashmap.end()) count += hashmap[curr_sum - k];

            if(hashmap.find(curr_sum) == hashmap.end()) hashmap[curr_sum] = 1;
            else hashmap[curr_sum] += 1;
        }

        return count;
    }
};

File: Array/560_subarray_sum_equals_k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        hashmap = {}
        prefix_sum = 0
        count = 0

        for num in nums:
            prefix_sum += num

            if prefix_sum == k:
                count += 1
            if prefix_sum - k in hashmap:
                count += hashmap[prefix_sum - k]

            hashmap[prefix_sum] = hashmap.get(prefix_sum, 0) + 1

        return count
        

# revised - 02/26/2025